core/src/main/scala/chisel3/Option.scala

```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3
+
+import chisel3.experimental.{BaseModule, SourceInfo}
+import chisel3.util.simpleClassName
+
+/**
+  * This object contains Chisel language definitions for describing configuration options and their accepted values.
+  */
+object option {
+
+  /** An option group declaration.
+    */
+  abstract class Option(implicit _sourceInfo: SourceInfo) {
+    self: Singleton =>
+
+    private[chisel3] def sourceInfo: SourceInfo = _sourceInfo
+
+    private[chisel3] def name: String = simpleClassName(this.getClass())
+
+    implicit def group: Option = this
+  }
+
+  /**
+    * An option case declaration.
+    */
+  abstract class Case(implicit val group: Option, _sourceInfo: SourceInfo) {
+    self: Singleton =>
+
+    private[chisel3] def sourceInfo: SourceInfo = _sourceInfo
+
+    private[chisel3] def name: String = simpleClassName(this.getClass())
+
+    /**
+      * A helper method to allow ModuleChoice to use the `->` syntax to specify case-module mappings.
+      *
+      * It captures a lazy reference to the module and produces a generator to avoid instantiating it.
+      *
+      * @param module Module to map to the current case.
+      */
+    def ->[T](module: => T): (Case, () => T) = (this, () => module)
```

Comment:
Re: the `ListMap` discussion at the end and guaranteeing uniqueness of arguments. There may be something clever that could be done to link the type of the `Option` into the case as a type parameter which could then be checked.